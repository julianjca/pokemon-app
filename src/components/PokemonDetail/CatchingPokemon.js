import React from "react";
import Image from "next/image";
import styled from "@emotion/styled";
import { keyframes } from "@emotion/react";
import Heading from "../Heading";
import { Spacing } from "../Shared";

const shake = keyframes`
  20% {
    transform: rotate(7deg);
  }
  
  40% {
    transform: rotate(-14deg);
  }
  
  60% {
    transform: rotate(4deg);
  }
  
  80% {
    transform: rotate(-2deg);
  }
  
  100% {
    transform: rotate(0deg);
  }
`;

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
`;

const PokeballWrapper = styled.div`
  animation: ${shake} 1s infinite ease-in-out;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const CatchingPokemon = () => {
  return (
    <Wrapper>
      <Spacing margin="0 0 20px 0">
        <Heading size="xl">Catching pokemon</Heading>
      </Spacing>
      <PokeballWrapper>
        <Image
          src="/pokeball.png"
          alt="catching pokemon"
          height={60}
          width={60}
        />
      </PokeballWrapper>
    </Wrapper>
  );
};

export default CatchingPokemon;
