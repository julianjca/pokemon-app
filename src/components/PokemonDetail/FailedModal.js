import React from "react";

import Heading from "../Heading";
import { Spacing } from "../Shared";
import Button from "../Button";

const SuccessModal = ({ onClick }) => {
  return (
    <div>
      <Heading size="xl">Oh no!</Heading>
      <Heading
        as="h3"
        size="md"
        style={{
          fontWeight: 500,
          marginTop: "10px",
        }}
      >
        You have failed to catch this pokemon.
      </Heading>
      <Spacing margin="20px 0 0">
        <Button onClick={onClick} fullWidth>
          Try Again!
        </Button>
      </Spacing>
    </div>
  );
};

export default SuccessModal;
