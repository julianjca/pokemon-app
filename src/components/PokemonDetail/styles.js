import styled from "@emotion/styled";

export const ImageWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const PokemonImage = styled.img`
  width: 45%;
  height: auto;

  @media (min-width: 720px) {
    width: 150px;
    height: 150px;
  }
`;

export const PokemonName = styled.h1`
  font-size: 2rem;
  text-align: center;
  text-transform: capitalize;
`;
