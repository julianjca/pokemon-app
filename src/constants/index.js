export const ADD_POKEMON = "ADD_POKEMON";
export const RELEASE_POKEMON = "RELEASE_POKEMON";
export const SET_POKEMONS = "SET_POKEMONS";
