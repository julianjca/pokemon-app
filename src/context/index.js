import React, { useEffect, createContext, useReducer } from "react";

const PokemonStateContext = createContext();
const PokemonDispatchContext = createContext();

const reducer = (state, action) => {
  switch (action.type) {
    case "ADD_POKEMON": {
      const pokemons = [...state.pokemons].concat(action.pokemon);
      localStorage.setItem("pokemons", JSON.stringify(pokemons));
      return { pokemons };
    }
    case "REMOVE_POKEMON": {
      return { pokemons: state.count - 1 };
    }
    case "SET_POKEMONS": {
      return { pokemons: action.pokemons };
    }
    default:
      console.log("error");
  }
};

const PokemonStateProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, {
    pokemons: [],
  });

  useEffect(() => {
    if (typeof window !== undefined) {
      const pokemons = localStorage.getItem("pokemons");
      if (pokemons) {
        dispatch({
          type: "SET_POKEMONS",
          pokemons: JSON.parse(pokemons),
        });
      }
    }
  }, []);

  return (
    <PokemonStateContext.Provider value={state}>
      <PokemonDispatchContext.Provider value={dispatch}>
        {children}
      </PokemonDispatchContext.Provider>
    </PokemonStateContext.Provider>
  );
};

const usePokemonState = () => {
  const context = React.useContext(PokemonStateContext);

  return context;
};

const usePokemonDispatch = () => {
  const context = React.useContext(PokemonDispatchContext);

  return context;
};

export { PokemonStateProvider, usePokemonState, usePokemonDispatch };
