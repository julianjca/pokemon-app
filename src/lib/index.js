export const makeSentenceCase = (str) => {
  return str
    .split(" ")
    .map((w) => w[0].toUpperCase() + w.substr(1).toLowerCase())
    .join(" ");
};

export const countPokemon = (pokemons, name) => {
  return pokemons.reduce(
    (acc, value) => ({
      ...acc,
      [value.pokemonName]: (acc[value.pokemonName] || 0) + 1,
    }),
    {}
  )[name];
};

export const uniqBy = (arr, key) => {
  const seen = {};

  return arr.filter((item) => {
    const k = key(item);
    return seen.hasOwnProperty(k) ? false : (seen[k] = true);
  });
};

export const getCatchingOdds = () => {
  return getCatchingOdds;
};
