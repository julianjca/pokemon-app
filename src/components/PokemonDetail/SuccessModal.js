import React, { useState, useCallback } from "react";

import Heading from "../Heading";
import Input from "../Input";
import { Spacing } from "../Shared";
import Button from "../Button";
import { usePokemonState } from "../../context";
import { ErrorMessage } from "./styles";

const SuccessModal = ({ onClick, pokemonName }) => {
  const [pokemonNickname, setPokemonNickname] = useState("");
  const [errorMessage, setErrorMessage] = useState(null);
  const state = usePokemonState();

  const handleSubmit = useCallback(() => {
    const pokemonIndex = state.pokemons.findIndex(
      (pokemon) =>
        pokemon.pokemonName === pokemonName &&
        pokemonNickname === pokemon.pokemonNickname
    );

    if (pokemonIndex === -1) {
      onClick(pokemonNickname);
      setErrorMessage(null);
      return;
    }

    setErrorMessage("Nickname should be unique.");
  }, [onClick, pokemonNickname, pokemonName, state.pokemons]);

  return (
    <div>
      <Heading size="xl">Congratulations!</Heading>
      <Heading
        as="h3"
        size="md"
        style={{
          fontWeight: 500,
          marginTop: "10px",
        }}
      >
        You have successfully catched this pokemon.
      </Heading>
      <Heading
        as="p"
        size="sm"
        style={{
          marginTop: "10px",
          fontWeight: 400,
        }}
      >
        Give it a name:
      </Heading>
      <Spacing margin="20px 0 0">
        <Input
          value={pokemonNickname}
          onChange={(e) => setPokemonNickname(e.target.value)}
          fullWidth
          type="text"
        />
        {errorMessage && <ErrorMessage>{errorMessage}</ErrorMessage>}
      </Spacing>
      <Spacing margin="20px 0 0">
        <Button onClick={handleSubmit} fullWidth>
          Save!
        </Button>
      </Spacing>
    </div>
  );
};

export default SuccessModal;
