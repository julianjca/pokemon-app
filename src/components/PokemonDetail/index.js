import React from "react";
import Image from "next/image";
import { Radar } from "react-chartjs-2";

import {
  ImageWrapper,
  PokemonImage,
  PokemonName,
  CatchPokemonButton,
  AttributesGrid,
  AttributesTitle,
  Tag,
  RadarChartWrapper,
  PokemonDetailContainer,
  TagsContainer,
  Left,
} from "./styles";
import { usePokemonDispatch, usePokemonState } from "../../context";
import Button from "../Button";
import { makeSentenceCase, countPokemon } from "../../lib";
import SuccessModal from "./SuccessModal";
import FailedModal from "./FailedModal";
import CatchingPokemon from "./CatchingPokemon";

const PokemonDetail = ({ pokemonData, handleOpenModal, handleCloseModal }) => {
  const dispatch = usePokemonDispatch();
  const state = usePokemonState();

  // eslint-disable-next-line no-unused-vars
  const pokemonCount = countPokemon(state.pokemons, pokemonData.name) || 0;

  const labels = [...pokemonData.stats].map(({ stat }) =>
    stat.name === "special-defense"
      ? "S. Defense"
      : makeSentenceCase(stat.name.split("-").join(" "))
  );
  const data = [...pokemonData.stats].map(({ base_stat }) => base_stat);

  const submitPokemon = (pokemonNickname) => {
    dispatch({
      type: "ADD_POKEMON",
      pokemon: {
        pokemonName: pokemonData.name,
        pokemonNickname,
        sprites: pokemonData.sprites,
      },
    });

    handleCloseModal();
  };

  const handleCatch = () => {
    // generate number 0 or 1 (50% chance to catch)
    const isCatched = Boolean(Math.round(Math.random()));

    // showing loading modal
    handleOpenModal(true, <CatchingPokemon />);

    if (isCatched) {
      setTimeout(() => {
        handleOpenModal(
          true,
          <SuccessModal
            onClick={submitPokemon}
            pokemonName={pokemonData.name}
          />
        );
      }, 1500);
      return;
    }
    setTimeout(() => {
      handleOpenModal(true, <FailedModal onClick={handleCatch} />);
    }, 1500);
  };

  return (
    <PokemonDetailContainer>
      <ImageWrapper>
        <PokemonImage>
          <img src={pokemonData.sprites.front_default} alt={pokemonData.name} />
        </PokemonImage>
        <PokemonImage>
          <img src={pokemonData.sprites.back_default} alt={pokemonData.name} />
        </PokemonImage>
      </ImageWrapper>
      <PokemonName>{pokemonData.name}</PokemonName>

      <AttributesGrid>
        <Left>
          <div>
            <AttributesTitle>Moves</AttributesTitle>
            <TagsContainer>
              {pokemonData.abilities.map(({ ability: { name } }) => (
                <Tag key={name}>{name}</Tag>
              ))}
            </TagsContainer>
          </div>
          <div>
            <AttributesTitle>Types</AttributesTitle>
            <TagsContainer>
              {pokemonData.types.map(({ type: { name } }) => (
                <Tag key={name}>{name}</Tag>
              ))}
            </TagsContainer>
          </div>
        </Left>
        <RadarChartWrapper>
          <Radar
            data={{
              datasets: [
                {
                  label: pokemonData.name,
                  backgroundColor: "#152C5740",
                  pointBackgroundColor: "#152C57",
                  borderColor: "#152C57",
                  data,
                },
              ],
              labels,
            }}
            height={50}
            width={50}
            options={{
              responsive: true,
              maintainAspectRatio: true,
            }}
          />
        </RadarChartWrapper>
      </AttributesGrid>

      <CatchPokemonButton onClick={handleCatch}>
        <Button
          style={{
            borderRadius: "100px",
            width: "220px",
            boxShadow: `0 10px 15px -3px rgba(0, 0, 0, 0.1),
            0 4px 6px -2px rgba(0, 0, 0, 0.05)`,
          }}
        >
          <div
            style={{
              display: "flex",
              alignItems: "center",
            }}
          >
            <Image
              src="/catch_pokeball.png"
              alt="pokeball"
              height={40}
              width={40}
            />
            <span
              style={{
                marginLeft: "20px",
              }}
            >
              Catch pokemon
            </span>
          </div>
        </Button>
      </CatchPokemonButton>
    </PokemonDetailContainer>
  );
};

export default PokemonDetail;
