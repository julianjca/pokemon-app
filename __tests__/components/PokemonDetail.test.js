/* eslint-disable testing-library/prefer-screen-queries */
import React from "react";
import { matchers } from "@emotion/jest";

import { render, fireEvent, act } from "../../test-utils";
import PokemonDetail from "../../src/components/PokemonDetail";
import * as lib from "../../src/lib";

expect.extend(matchers);

HTMLCanvasElement.prototype.getContext = () => {
  // return whatever getContext has to return
};

jest.mock("react-chartjs-2", () => ({
  Radar: () => null,
}));

jest.mock("../../src/lib");

const pokemonData = {
  __typename: "Pokemon",
  name: "bulbasaur",
  sprites: {
    __typename: "Sprite",
    front_default:
      "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
    back_default:
      "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/1.png",
  },
  abilities: [
    {
      __typename: "Ability",
      ability: {
        __typename: "BaseName",
        url: "https://pokeapi.co/api/v2/ability/65/",
        name: "overgrow",
      },
    },
    {
      __typename: "Ability",
      ability: {
        __typename: "BaseName",
        url: "https://pokeapi.co/api/v2/ability/34/",
        name: "chlorophyll",
      },
    },
  ],
  stats: [
    {
      __typename: "Stat",
      base_stat: 45,
      effort: 0,
      stat: { __typename: "BaseName", name: "hp" },
    },
    {
      __typename: "Stat",
      base_stat: 49,
      effort: 0,
      stat: { __typename: "BaseName", name: "attack" },
    },
    {
      __typename: "Stat",
      base_stat: 49,
      effort: 0,
      stat: { __typename: "BaseName", name: "defense" },
    },
    {
      __typename: "Stat",
      base_stat: 65,
      effort: 1,
      stat: { __typename: "BaseName", name: "special-attack" },
    },
    {
      __typename: "Stat",
      base_stat: 65,
      effort: 0,
      stat: { __typename: "BaseName", name: "special-defense" },
    },
    {
      __typename: "Stat",
      base_stat: 45,
      effort: 0,
      stat: { __typename: "BaseName", name: "speed" },
    },
  ],
  types: [
    {
      __typename: "Type",
      type: {
        __typename: "BaseName",
        name: "grass",
        url: "https://pokeapi.co/api/v2/type/12/",
      },
    },
    {
      __typename: "Type",
      type: {
        __typename: "BaseName",
        name: "poison",
        url: "https://pokeapi.co/api/v2/type/4/",
      },
    },
  ],
};

describe("PokemonDetail", () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });
  test("renders the component", () => {
    const element = render(<PokemonDetail pokemonData={pokemonData} />);

    expect(element).toMatchSnapshot();
  });

  test("renders pokemon data", () => {
    const { getByText, getByAltText } = render(
      <PokemonDetail pokemonData={pokemonData} />
    );

    expect(getByText("bulbasaur")).toBeInTheDocument();
    expect(getByText("overgrow")).toBeInTheDocument();
    expect(getByText("chlorophyll")).toBeInTheDocument();
    expect(getByText("grass")).toBeInTheDocument();
    expect(getByText("poison")).toBeInTheDocument();
    expect(getByAltText("bulbasaur_front")).toBeInTheDocument();
    expect(getByAltText("bulbasaur_back")).toBeInTheDocument();
  });

  test("renders catch pokemon button and showing success modal", () => {
    const handleOpenModal = jest.fn();
    const catchingOdds = jest
      .spyOn(lib, "getCatchingOdds")
      .mockReturnValueOnce(true);

    const { getByTestId } = render(
      <PokemonDetail
        pokemonData={pokemonData}
        handleOpenModal={handleOpenModal}
      />
    );

    const catchPokemonButton = getByTestId("catch_pokemon");
    expect(catchPokemonButton).toBeInTheDocument();

    act(() => {
      fireEvent.click(catchPokemonButton);
      jest.runAllTimers();
      // runs the first loading modal
      // open success modal form
      expect(handleOpenModal).toHaveBeenCalledTimes(2);
    });

    catchingOdds.mockRestore();
  });

  test("renders catch pokemon button and showing failed modal", () => {
    const handleOpenModal = jest.fn();
    const catchingOdds = jest
      .spyOn(lib, "getCatchingOdds")
      .mockReturnValueOnce(false);

    const { getByTestId } = render(
      <PokemonDetail
        pokemonData={pokemonData}
        handleOpenModal={handleOpenModal}
      />
    );

    const catchPokemonButton = getByTestId("catch_pokemon");
    expect(catchPokemonButton).toBeInTheDocument();

    act(() => {
      fireEvent.click(catchPokemonButton);

      jest.runAllTimers();

      // runs the first loading modal
      // open failed modal form
      expect(handleOpenModal).toHaveBeenCalledTimes(2);
    });

    catchingOdds.mockRestore();
  });
});
