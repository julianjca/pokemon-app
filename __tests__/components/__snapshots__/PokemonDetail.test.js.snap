// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonDetail renders the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .emotion-0 {
  padding: 4rem 0 8rem;
}

@media (min-width: 720px) {
  .emotion-0 {
    padding: 4rem 0;
  }
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-4 img {
  width: 150px;
  height: 150px;
}

@media (min-width: 720px) {
  .emotion-4 img {
    width: 300px;
    height: 300px;
  }
}

.emotion-8 {
  width: 100%;
  max-width: 1000px;
  margin: 40px auto 0;
}

@media (min-width: 1000px) {
  .emotion-8 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    justify-content: space-between;
  }
}

.emotion-10 {
  font-size: 3rem;
  text-align: left;
  text-transform: capitalize;
  font-weight: bold;
  color: #193569;
}

@media (min-width: 720px) {
  .emotion-14 {
    width: 45%;
  }
}

.emotion-16 {
  font-weight: 600;
  font-size: 1.2rem;
}

.emotion-18 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-top: 1rem;
}

.emotion-20 {
  box-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1),0 4px 6px -2px rgba(0, 0, 0, 0.05);
  padding: 12px 15px;
  border-radius: 100px;
  margin-right: 10px;
  margin-bottom: 10px;
  color: #193569;
  background: #FDC903;
  text-transform: capitalize;
  font-weight: 700;
}

.emotion-32 {
  width: 100%;
  margin-top: 60px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
}

@media (min-width: 720px) {
  .emotion-32 {
    width: 90%;
    margin-top: 0;
  }
}

.emotion-34 {
  position: fixed;
  left: 50%;
  bottom: 40px;
  -webkit-transform: translateX(-50%);
  -moz-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.emotion-36 {
  background-color: #193569;
  border-radius: 0.5em;
  border: none;
  padding: 1em 1.2em;
  font-size: 1rem;
  font-weight: bold;
  color: #FDC903;
  cursor: pointer;
  -webkit-transition: 0.15s all ease-in-out;
  transition: 0.15s all ease-in-out;
  outline: none;
}

.emotion-36:hover {
  background-color: #152c57;
  -webkit-transition: 0.15s all ease-in-out;
  transition: 0.15s all ease-in-out;
}

<body>
    <div>
      <div
        class="emotion-0 emotion-1"
      >
        <div
          class="emotion-2 emotion-3"
        >
          <div
            class="emotion-4 emotion-5"
          >
            <img
              alt="bulbasaur_front"
              src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
            />
          </div>
          <div
            class="emotion-4 emotion-5"
          >
            <img
              alt="bulbasaur_back"
              src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/1.png"
            />
          </div>
        </div>
        <div
          class="emotion-8 emotion-9"
        >
          <h1
            class="emotion-10 emotion-11"
          >
            bulbasaur
          </h1>
        </div>
        <div
          class="emotion-8 emotion-9"
        >
          <div
            class="emotion-14 emotion-15"
          >
            <div>
              <h4
                class="emotion-16 emotion-17"
              >
                Moves
              </h4>
              <div
                class="emotion-18 emotion-19"
              >
                <div
                  class="emotion-20 emotion-21"
                >
                  overgrow
                </div>
                <div
                  class="emotion-20 emotion-21"
                >
                  chlorophyll
                </div>
              </div>
            </div>
            <div>
              <h4
                class="emotion-16 emotion-17"
              >
                Types
              </h4>
              <div
                class="emotion-18 emotion-19"
              >
                <div
                  class="emotion-20 emotion-21"
                >
                  grass
                </div>
                <div
                  class="emotion-20 emotion-21"
                >
                  poison
                </div>
              </div>
            </div>
          </div>
          <div
            class="emotion-32 emotion-33"
          />
        </div>
        <div
          class="emotion-34 emotion-35"
          data-testid="catch_pokemon"
        >
          <button
            class="emotion-36 emotion-37"
            style="border-radius: 100px; width: 220px; box-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1),
            0 4px 6px -2px rgba(0, 0, 0, 0.05);"
          >
            <div
              style="display: flex; align-items: center;"
            >
              <div
                style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
              >
                <div
                  style="box-sizing: border-box; display: block; max-width: 100%;"
                >
                  <img
                    alt=""
                    aria-hidden="true"
                    role="presentation"
                    src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNDAiIGhlaWdodD0iNDAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                    style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                  />
                </div>
                <img
                  alt="pokeball"
                  decoding="async"
                  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                  style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                />
              </div>
              <span
                style="margin-left: 20px;"
              >
                Catch pokemon
              </span>
            </div>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": .emotion-0 {
  padding: 4rem 0 8rem;
}

@media (min-width: 720px) {
  .emotion-0 {
    padding: 4rem 0;
  }
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-4 img {
  width: 150px;
  height: 150px;
}

@media (min-width: 720px) {
  .emotion-4 img {
    width: 300px;
    height: 300px;
  }
}

.emotion-8 {
  width: 100%;
  max-width: 1000px;
  margin: 40px auto 0;
}

@media (min-width: 1000px) {
  .emotion-8 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    justify-content: space-between;
  }
}

.emotion-10 {
  font-size: 3rem;
  text-align: left;
  text-transform: capitalize;
  font-weight: bold;
  color: #193569;
}

@media (min-width: 720px) {
  .emotion-14 {
    width: 45%;
  }
}

.emotion-16 {
  font-weight: 600;
  font-size: 1.2rem;
}

.emotion-18 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-top: 1rem;
}

.emotion-20 {
  box-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1),0 4px 6px -2px rgba(0, 0, 0, 0.05);
  padding: 12px 15px;
  border-radius: 100px;
  margin-right: 10px;
  margin-bottom: 10px;
  color: #193569;
  background: #FDC903;
  text-transform: capitalize;
  font-weight: 700;
}

.emotion-32 {
  width: 100%;
  margin-top: 60px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
}

@media (min-width: 720px) {
  .emotion-32 {
    width: 90%;
    margin-top: 0;
  }
}

.emotion-34 {
  position: fixed;
  left: 50%;
  bottom: 40px;
  -webkit-transform: translateX(-50%);
  -moz-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.emotion-36 {
  background-color: #193569;
  border-radius: 0.5em;
  border: none;
  padding: 1em 1.2em;
  font-size: 1rem;
  font-weight: bold;
  color: #FDC903;
  cursor: pointer;
  -webkit-transition: 0.15s all ease-in-out;
  transition: 0.15s all ease-in-out;
  outline: none;
}

.emotion-36:hover {
  background-color: #152c57;
  -webkit-transition: 0.15s all ease-in-out;
  transition: 0.15s all ease-in-out;
}

<div>
    <div
      class="emotion-0 emotion-1"
    >
      <div
        class="emotion-2 emotion-3"
      >
        <div
          class="emotion-4 emotion-5"
        >
          <img
            alt="bulbasaur_front"
            src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
          />
        </div>
        <div
          class="emotion-4 emotion-5"
        >
          <img
            alt="bulbasaur_back"
            src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/1.png"
          />
        </div>
      </div>
      <div
        class="emotion-8 emotion-9"
      >
        <h1
          class="emotion-10 emotion-11"
        >
          bulbasaur
        </h1>
      </div>
      <div
        class="emotion-8 emotion-9"
      >
        <div
          class="emotion-14 emotion-15"
        >
          <div>
            <h4
              class="emotion-16 emotion-17"
            >
              Moves
            </h4>
            <div
              class="emotion-18 emotion-19"
            >
              <div
                class="emotion-20 emotion-21"
              >
                overgrow
              </div>
              <div
                class="emotion-20 emotion-21"
              >
                chlorophyll
              </div>
            </div>
          </div>
          <div>
            <h4
              class="emotion-16 emotion-17"
            >
              Types
            </h4>
            <div
              class="emotion-18 emotion-19"
            >
              <div
                class="emotion-20 emotion-21"
              >
                grass
              </div>
              <div
                class="emotion-20 emotion-21"
              >
                poison
              </div>
            </div>
          </div>
        </div>
        <div
          class="emotion-32 emotion-33"
        />
      </div>
      <div
        class="emotion-34 emotion-35"
        data-testid="catch_pokemon"
      >
        <button
          class="emotion-36 emotion-37"
          style="border-radius: 100px; width: 220px; box-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1),
            0 4px 6px -2px rgba(0, 0, 0, 0.05);"
        >
          <div
            style="display: flex; align-items: center;"
          >
            <div
              style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
            >
              <div
                style="box-sizing: border-box; display: block; max-width: 100%;"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  role="presentation"
                  src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNDAiIGhlaWdodD0iNDAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                  style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                />
              </div>
              <img
                alt="pokeball"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
            </div>
            <span
              style="margin-left: 20px;"
            >
              Catch pokemon
            </span>
          </div>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
